File: lib/symphony/repo_planner.ex
defmodule Symphony.RepoPlanner do
  @moduledoc false

  alias Symphony.CodingContext.{CodingClassification}
  alias Symphony.CodingContext
  alias Symphony.CodexClient

  alias Symphony.Config.{
    CodexConfig,
    CodingContextConfig,
    RepositoryConfig,
    RepositoryPlanningConfig
  }

  alias Symphony.Models.{Issue, RepoPlan, RepoPlanItem}
  alias Symphony.Utils

  def plan_repositories(
        %Issue{} = issue,
        %RepositoryPlanningConfig{} = config,
        %CodingContextConfig{} = coding_config,
        %CodingClassification{} = classification,
        opts \\ []
      ) do
    cond do
      !config.enabled ->
        nil

      !classification.is_coding_task ->
        %RepoPlan{
          issue_identifier: issue.identifier,
          coding_task: false,
          planner: config.planner,
          source: classification.source,
          confidence: classification.confidence,
          notes: "Issue was not classified as a coding task.",
          created_at: Utils.now_utc()
        }

      config.planner == "llm" ->
        try do
          plan_with_llm(issue, config, coding_config, opts)
        rescue
          error ->
            if config.fallback == "block" do
              %RepoPlan{
                issue_identifier: issue.identifier,
                coding_task: true,
                planner: config.planner,
                source: "fallback:block",
                needs_human: true,
                human_reason:
                  "Repository planner failed and fallback is block: #{Utils.truncate(Exception.message(error), 300)}",
                created_at: Utils.now_utc()
              }
            else
              plan_with_rules(issue, config, "fallback:rules")
            end
        end

      true ->
        plan_with_rules(issue, config, "rules")
    end
  end

  def apply_repo_plan_to_prompt(prompt, nil, _workspace_path), do: prompt

  def apply_repo_plan_to_prompt(prompt, %RepoPlan{coding_task: false}, _workspace_path),
    do: prompt

  def apply_repo_plan_to_prompt(prompt, %RepoPlan{} = repo_plan, workspace_path) do
    plan_json = Jason.encode!(RepoPlan.to_map(repo_plan), pretty: true)
    git_metadata = workspace_git_metadata(workspace_path)

    git_guardrail =
      "Git hygiene guardrail: commit and push only the expected branch recorded for each repo in `.symphony-workspace.json`. Never push an inherited source checkout branch. If `git branch --show-current` differs from the repo's expected branch, stop and report the mismatch. The workspace-local pre-push hook rejects pushes to any other branch.\n"

    git_guardrail =
      if git_metadata == [] do
        git_guardrail
      else
        git_guardrail <> "Prepared git branches:\n#{Jason.encode!(git_metadata, pretty: true)}\n"
      end

    """
    <symphony_repo_plan>
    Symphony prepared an explicit repository plan for this issue. Treat it as a guardrail, not as proof the implementation is already understood.

    Workspace root: #{workspace_path}
    Repositories are checked out under `repos/<path_name>` inside the workspace root.
    Start by inspecting the primary repo. You may read secondary and read-only context repos as needed. Only edit the primary repo and secondary repos whose `edit_allowed` value is true. Do not edit read-only context repos. If the current Linear issue text proves the repo plan is wrong or incomplete, stop and report that instead of patching an unapproved repo.

    #{git_guardrail}
    #{plan_json}
    </symphony_repo_plan>

    #{prompt}
    """
    |> String.trim()
  end

  defp workspace_git_metadata(workspace_path) do
    path = Path.join(workspace_path, ".symphony-workspace.json")

    with {:ok, body} <- File.read(path),
         {:ok, payload} when is_map(payload) <- Jason.decode(body),
         repositories when is_list(repositories) <- payload["repositories"] do
      repositories
      |> Enum.flat_map(fn
        item when is_map(item) ->
          if is_map(item["git"]) do
            [
              %{
                "slug" => item["slug"],
                "path" => item["path"],
                "edit_allowed" => item["edit_allowed"],
                "expected_branch" => item["git"]["expected_branch"],
                "expected_ref" => item["git"]["expected_ref"],
                "base_ref" => item["git"]["base_ref"]
              }
            ]
          else
            []
          end

        _ ->
          []
      end)
    else
      _ -> []
    end
  end

  defp plan_with_llm(issue, config, coding_config, opts) do
    %CodexConfig{} = codex_config = Keyword.fetch!(opts, :codex_config)
    workspace_path = Keyword.fetch!(opts, :workspace_path)
    codex_client = Keyword.get(opts, :codex_client, CodexClient)

    planner_codex_config = %{
      codex_config
      | model: config.plan_model || codex_config.model,
        effort: config.plan_effort,
        turn_timeout_ms: config.plan_timeout_ms,
        summary: nil,
        personality: nil
    }

    session =
      apply(codex_client, :start_session, [
        planner_codex_config,
        workspace_path,
        [tracker_config: nil, on_event: fn _event -> :ok end]
      ])

    try do
      {result, _session} =
        apply(codex_client, :run_turn, [
          session,
          planner_prompt(issue, config, coding_config),
          [capture_agent_text: true]
        ])

      result
      |> agent_message_text()
      |> parse_json_object()
      |> normalize_plan(issue, config, planner: "llm", source: "llm")
      |> apply_rules_crosscheck(issue, config)
    after
      apply(codex_client, :stop_session, [session])
    end
  end

  defp planner_prompt(issue, config, coding_config) do
    payload = %{
      "issue" => Issue.to_template_data(issue),
      "repositories" => Enum.map(config.repositories, &RepositoryConfig.to_prompt_data/1),
      "coding_context" =>
        if(coding_config.enabled,
          do: coding_config |> CodingContext.load_coding_context() |> Utils.truncate(30_000),
          else: ""
        )
    }

    """
    You are Symphony's repository planner for a background coding agent.
    Pick the repository set the agent is allowed to use for this Linear issue. Many valid tasks span multiple repos, so return a primary repo plus optional secondary/edit repos and read-only context repos. Prefer the runtime/product repo that owns the behavior as primary. Do not choose a gateway/config repo merely because the issue mentions AI, search, prompts, or suggestions if the runtime/UI/provider adapter lives elsewhere. If the issue requests a specific provider/integration, keep that provider direction in the reason.

    Return only one JSON object, no markdown, no prose, and no tool calls.
    Schema:
    {
      "coding_task": true,
      "primary_repo": {"slug": "owner/name", "reason": "why this repo is the start"},
      "secondary_repos": [{"slug": "owner/name", "reason": "why it may need edits", "edit_allowed": true}],
      "read_only_context_repos": [{"slug": "owner/name", "reason": "why it is useful context"}],
      "confidence": 0.0,
      "needs_human": false,
      "human_reason": null,
      "notes": "short operational note"
    }

    Rules:
    - Use only repository slugs listed in the input catalog.
    - If there is no clear primary repo, set needs_human=true and explain.
    - If a secondary repo might need edits, include it as secondary_repos with edit_allowed=true.
    - If a repo is only background material, include it as read_only_context_repos.
    - If the issue is not a coding/repository task, set coding_task=false and leave repo lists empty.
    - Route post-call, saved-call, call-history, history-detail, history-tab, recap, follow-up email, and template UI work to the web/customer app repository when one exists. Do not route those issues to a desktop/live-runtime repository unless the issue explicitly names desktop overlay, native capture, transcription, or live in-call behavior.

    Planner input JSON:
    #{Jason.encode!(payload)}
    """
  end

  defp plan_with_rules(issue, config, source) do
    text =
      "#{issue.identifier}\n#{issue.title}\n#{issue.description || ""}\n#{Enum.join(issue.labels, " ")}"
      |> String.downcase()

    scored =
      config.repositories
      |> Enum.flat_map(fn repo ->
        {score, strong_score, reasons} = score_repo(repo, text)
        if score > 0, do: [{score, strong_score, repo, Enum.take(reasons, 6)}], else: []
      end)
      |> Enum.sort_by(fn {score, strong_score, repo, _reasons} ->
        {-score, -strong_score, repo.slug}
      end)

    case scored do
      [] ->
        %RepoPlan{
          issue_identifier: issue.identifier,
          coding_task: true,
          planner: config.planner,
          source: source,
          needs_human: true,
          human_reason: "No configured repository matched the issue text.",
          confidence: 0.0,
          created_at: Utils.now_utc()
        }

      [{top_score, top_strong_score, top_repo, top_reasons} | rest] ->
        tied =
          scored
          |> Enum.filter(fn {score, strong_score, _repo, _} ->
            score == top_score and strong_score == top_strong_score
          end)
          |> Enum.map(fn {_, _, repo, _} -> repo.slug end)

        needs_human = length(tied) > 1

        %RepoPlan{
          issue_identifier: issue.identifier,
          coding_task: true,
          planner: config.planner,
          source: source,
          primary_repo:
            item(top_repo, "primary", "Rules matched: #{Enum.join(top_reasons, ", ")}"),
          secondary_repos:
            rest
            |> Enum.take(3)
            |> Enum.map(fn {_score, _strong_score, repo, reasons} ->
              item(repo, "secondary", "Rules also matched: #{Enum.join(reasons, ", ")}")
            end),
          confidence: min(0.85, max(0.2, top_score / 12)),
          needs_human: needs_human,
          human_reason:
            if(needs_human,
              do: "Rules planner found tied primary repositories: #{Enum.join(tied, ", ")}"
            ),
          created_at: Utils.now_utc()
        }
    end
  end

  defp apply_rules_crosscheck(
         %RepoPlan{coding_task: true, primary_repo: %RepoPlanItem{} = llm_primary} = plan,
         %Issue{} = issue,
         %RepositoryPlanningConfig{} = config
       ) do
    rules_plan = plan_with_rules(issue, config, "llm:rules_crosscheck")
    rules_primary = rules_plan.primary_repo

    cond do
      rules_plan.needs_human or is_nil(rules_primary) ->
        plan

      rules_primary.slug == llm_primary.slug ->
        plan

      not Enum.any?(plan.read_only_context_repos, &(&1.slug == rules_primary.slug)) ->
        plan

      true ->
        promote_rules_primary(plan, rules_primary, llm_primary)
    end
  end

  defp apply_rules_crosscheck(plan, _issue, _config), do: plan

  defp promote_rules_primary(%RepoPlan{} = plan, %RepoPlanItem{} = rules_primary, llm_primary) do
    promoted =
      rules_primary
      |> retag_plan_item("primary", true)
      |> Map.put(
        :reason,
        "#{rules_primary.reason}; promoted over LLM primary #{llm_primary.slug} because the LLM marked this rules-matched repo as read-only context."
      )

    demoted =
      llm_primary
      |> retag_plan_item("read_only_context", false)
      |> Map.put(
        :reason,
        "LLM initially selected this as primary, but rules cross-check promoted #{rules_primary.slug}."
      )

    secondary =
      plan.secondary_repos
      |> Enum.reject(&(&1.slug in [rules_primary.slug, llm_primary.slug]))

    read_only =
      [demoted | plan.read_only_context_repos]
      |> Enum.reject(&(&1.slug == rules_primary.slug))
      |> dedupe_items()

    %{
      plan
      | primary_repo: promoted,
        secondary_repos: secondary,
        read_only_context_repos: read_only,
        source: "llm+rules_crosscheck",
        notes:
          [
            plan.notes,
            "Rules cross-check promoted #{rules_primary.slug} over #{llm_primary.slug}."
          ]
          |> Enum.reject(&blank?/1)
          |> Enum.join(" ")
    }
  end

  defp retag_plan_item(%RepoPlanItem{} = item, role, edit_allowed) do
    %{item | role: role, edit_allowed: edit_allowed}
  end

  defp blank?(value), do: is_nil(value) or String.trim(to_string(value)) == ""

  defp score_repo(%RepositoryConfig{} = repo, text) do
    candidates =
      [
        {String.downcase(repo.slug), 4},
        {String.downcase(RepositoryConfig.path_name(repo)), 4}
        | Enum.map(repo.aliases, &{String.downcase(&1), 3})
      ]
      |> Enum.uniq_by(fn {candidate, _weight} -> candidate end)

    {score, strong_score, reasons} =
      candidates
      |> Enum.reject(fn {candidate, _weight} -> candidate == "" end)
      |> Enum.reduce({0, 0, []}, fn {candidate, weight}, {score, strong_score, reasons} ->
        if term_matches?(text, candidate) do
          {score + weight, strong_score + weight, append_reason(reasons, candidate)}
        else
          {score, strong_score, reasons}
        end
      end)

    if repo.description do
      repo.description
      |> keyword_terms()
      |> Enum.reduce({score, strong_score, reasons}, fn word, {score, strong_score, reasons} ->
        if term_matches?(text, word),
          do: {score + 1, strong_score, append_reason(reasons, word)},
          else: {score, strong_score, reasons}
      end)
    else
      {score, strong_score, reasons}
    end
  end

  defp term_matches?(_text, ""), do: false

  defp term_matches?(text, term) do
    term = String.trim(term)

    if term == "" do
      false
    else
      pattern = "(^|[^[:alnum:]])" <> Regex.escape(term) <> "($|[^[:alnum:]])"
      Regex.match?(Regex.compile!(pattern, "u"), text)
    end
  end

  defp append_reason(reasons, reason) do
    if reason in reasons, do: reasons, else: reasons ++ [reason]
  end

  defp item(%RepositoryConfig{} = repo, role, reason, opts \\ []) do
    edit_allowed =
      Keyword.get(opts, :edit_allowed, role != "read_only_context") and
        role != "read_only_context"

    %RepoPlanItem{
      slug: repo.slug,
      role: role,
      reason: reason,
      path_name: Utils.sanitize_workspace_key(RepositoryConfig.path_name(repo)),
      edit_allowed: edit_allowed
    }
  end

  defp normalize_plan(data, issue, config, opts) do
    known = RepositoryPlanningConfig.repository_by_slug(config)
    unknown = []

    {primary, unknown} =
      parse_plan_item(Utils.map_get(data, "primary_repo"), "primary", known, unknown)

    {secondary, unknown} =
      data
      |> Utils.map_get("secondary_repos")
      |> list_value()
      |> Enum.reduce({[], unknown}, fn raw, {items, unknown_acc} ->
        case parse_plan_item(raw, "secondary", known, unknown_acc) do
          {nil, next_unknown} -> {items, next_unknown}
          {item, next_unknown} -> {items ++ [item], next_unknown}
        end
      end)

    {read_only, unknown} =
      data
      |> Utils.map_get("read_only_context_repos")
      |> list_value()
      |> Enum.reduce({[], unknown}, fn raw, {items, unknown_acc} ->
        case parse_plan_item(raw, "read_only_context", known, unknown_acc) do
          {nil, next_unknown} -> {items, next_unknown}
          {item, next_unknown} -> {items ++ [item], next_unknown}
        end
      end)

    secondary = dedupe_items(secondary)
    read_only = dedupe_items(read_only)

    secondary =
      if primary,
        do: Enum.reject(secondary, &(&1.slug == primary.slug)),
        else: secondary

    read_only =
      if primary,
        do: Enum.reject(read_only, &(&1.slug == primary.slug)),
        else: read_only

    coding_task =
      truthy?(Utils.map_get(data, "coding_task", Utils.map_get(data, "is_coding_task", true)))

    needs_human = truthy?(Utils.map_get(data, "needs_human"))
    human_reason = clean_truncated(Utils.map_get(data, "human_reason"), 500)

    {needs_human, human_reason} =
      if coding_task and is_nil(primary) do
        {true, human_reason || "Repository planner did not return a primary repo."}
      else
        {needs_human, human_reason}
      end

    {needs_human, human_reason} =
      if unknown != [] do
        suffix =
          "Planner returned unknown repositories: #{unknown |> Enum.uniq() |> Enum.sort() |> Enum.join(", ")}."

        {true,
         [human_reason, suffix] |> Enum.reject(&is_nil/1) |> Enum.join(" ") |> String.trim()}
      else
        {needs_human, human_reason}
      end

    %RepoPlan{
      issue_identifier: issue.identifier,
      coding_task: coding_task,
      planner: Keyword.fetch!(opts, :planner),
      source: Keyword.fetch!(opts, :source),
      primary_repo: primary,
      secondary_repos: secondary,
      read_only_context_repos: read_only,
      confidence: confidence(Utils.map_get(data, "confidence")),
      needs_human: needs_human,
      human_reason: if(human_reason == "", do: nil, else: human_reason),
      notes: clean_truncated(Utils.map_get(data, "notes"), 500),
      created_at: Utils.now_utc()
    }
  end

  defp parse_plan_item(raw, role, known, unknown) when is_map(raw) do
    slug = raw |> Utils.map_get("slug") |> to_string() |> String.trim()

    cond do
      slug == "" ->
        {nil, unknown}

      repo = known[slug] ->
        edit_allowed = truthy?(Utils.map_get(raw, "edit_allowed", role != "read_only_context"))

        {item(repo, role, clean_truncated(Utils.map_get(raw, "reason"), 500),
           edit_allowed: edit_allowed
         ), unknown}

      true ->
        {nil, unknown ++ [slug]}
    end
  end

  defp parse_plan_item(_raw, _role, _known, unknown), do: {nil, unknown}

  defp dedupe_items(items) do
    {deduped, _seen} =
      Enum.reduce(items, {[], MapSet.new()}, fn item, {acc, seen} ->
        if MapSet.member?(seen, item.slug),
          do: {acc, seen},
          else: {acc ++ [item], MapSet.put(seen, item.slug)}
      end)

    deduped
  end

  defp list_value(value) when is_list(value), do: value
  defp list_value(_value), do: []

  defp confidence(value) do
    case Utils.to_float(value) do
      nil -> nil
      float -> float |> max(0.0) |> min(1.0)
    end
  end

  defp truthy?(false), do: false
  defp truthy?(nil), do: false
  defp truthy?(_), do: true

  defp clean_truncated(nil, _limit), do: nil

  defp clean_truncated(value, limit) do
    text = value |> to_string() |> Utils.truncate(limit) |> String.trim()
    if text == "", do: nil, else: text
  end

  defp agent_message_text(result) when is_map(result) do
    Map.get(result, :agent_message_text) || Map.get(result, "agent_message_text") || ""
  end

  defp agent_message_text(_), do: ""

  defp parse_json_object(text) do
    stripped = String.trim(to_string(text || ""))

    if stripped == "" do
      raise ArgumentError, message: "repo planner returned empty text"
    end

    value =
      case Jason.decode(stripped) do
        {:ok, decoded} ->
          decoded

        {:error, _reason} ->
          stripped |> extract_json_object() |> Jason.decode!()
      end

    unless is_map(value) do
      raise ArgumentError, message: "repo planner JSON is not an object"
    end

    value
  end

  defp extract_json_object(text) do
    start = :binary.match(text, "{")
    finish = :binary.matches(text, "}") |> List.last()

    case {start, finish} do
      {{start_index, 1}, {end_index, 1}} when end_index > start_index ->
        binary_part(text, start_index, end_index - start_index + 1)

      _ ->
        raise ArgumentError, message: "repo planner output did not contain a JSON object"
    end
  end

  defp keyword_terms(text) do
    stop =
      MapSet.new([
        "the",
        "and",
        "for",
        "with",
        "from",
        "that",
        "this",
        "repo",
        "choose",
        "validation",
        "local",
        "path"
      ])

    text
    |> String.downcase()
    |> String.replace(~r/[\/-]/, " ")
    |> String.split()
    |> Enum.map(&String.replace(&1, ~r/[^[:alnum:]]/, ""))
    |> Enum.filter(&(String.length(&1) >= 5 and !MapSet.member?(stop, &1)))
    |> Enum.take(40)
  end
end

File: symphony/repo_planner.py
from __future__ import annotations

import json
from dataclasses import replace
from pathlib import Path
from typing import Any

from .codex_client import CodexAppServerSession
from .coding_context import CodingClassification, load_coding_context
from .config import CodexConfig, CodingContextConfig, RepositoryConfig, RepositoryPlanningConfig
from .models import Issue, RepoPlan, RepoPlanItem
from .utils import now_utc, sanitize_workspace_key, truncate


async def plan_repositories(
    issue: Issue,
    config: RepositoryPlanningConfig,
    coding_config: CodingContextConfig,
    classification: CodingClassification,
    *,
    codex_config: CodexConfig,
    workspace_path: Path,
) -> RepoPlan | None:
    if not config.enabled:
        return None
    if not classification.is_coding_task:
        return RepoPlan(
            issue_identifier=issue.identifier,
            coding_task=False,
            planner=config.planner,
            source=classification.source,
            confidence=classification.confidence,
            notes="Issue was not classified as a coding task.",
            created_at=now_utc(),
        )
    if config.planner == "llm":
        try:
            return await _plan_with_llm(issue, config, coding_config, codex_config, workspace_path)
        except Exception as exc:
            if config.fallback == "block":
                return RepoPlan(
                    issue_identifier=issue.identifier,
                    coding_task=True,
                    planner=config.planner,
                    source="fallback:block",
                    needs_human=True,
                    human_reason=f"Repository planner failed and fallback is block: {truncate(str(exc), 300)}",
                    created_at=now_utc(),
                )
    return _plan_with_rules(issue, config, source="rules" if config.planner == "rules" else "fallback:rules")


def apply_repo_plan_to_prompt(prompt: str, repo_plan: RepoPlan | None, workspace_path: Path) -> str:
    if repo_plan is None or not repo_plan.coding_task:
        return prompt
    plan_json = json.dumps(repo_plan.to_dict(), sort_keys=True, default=str, indent=2)
    git_metadata = _workspace_git_metadata(workspace_path)
    git_guardrail = (
        "Git hygiene guardrail: commit and push only the expected branch recorded for each repo in "
        "`.symphony-workspace.json`. Never push an inherited source checkout branch. If "
        "`git branch --show-current` differs from the repo's expected branch, stop and report the mismatch. "
        "The workspace-local pre-push hook rejects pushes to any other branch.\n"
    )
    if git_metadata:
        git_guardrail += f"Prepared git branches:\n{json.dumps(git_metadata, sort_keys=True, default=str, indent=2)}\n"
    return (
        "<symphony_repo_plan>\n"
        "Symphony prepared an explicit repository plan for this issue. Treat it as a guardrail, not as proof the "
        "implementation is already understood.\n\n"
        f"Workspace root: {workspace_path}\n"
        "Repositories are checked out under `repos/<path_name>` inside the workspace root.\n"
        "Start by inspecting the primary repo. You may read secondary and read-only context repos as needed. "
        "Only edit the primary repo and secondary repos whose `edit_allowed` value is true. Do not edit "
        "read-only context repos. If the current Linear issue text proves the repo plan is wrong or incomplete, "
        "stop and report that instead of patching an unapproved repo.\n\n"
        f"{git_guardrail}\n"
        f"{plan_json}\n"
        "</symphony_repo_plan>\n\n"
        f"{prompt}"
    )


def _workspace_git_metadata(workspace_path: Path) -> list[dict[str, Any]]:
    metadata_path = workspace_path / ".symphony-workspace.json"
    if not metadata_path.exists():
        return []
    try:
        payload = json.loads(metadata_path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return []
    repositories = payload.get("repositories")
    if not isinstance(repositories, list):
        return []
    result: list[dict[str, Any]] = []
    for item in repositories:
        if not isinstance(item, dict):
            continue
        git = item.get("git")
        if not isinstance(git, dict):
            continue
        result.append(
            {
                "slug": item.get("slug"),
                "path": item.get("path"),
                "edit_allowed": item.get("edit_allowed"),
                "expected_branch": git.get("expected_branch"),
                "expected_ref": git.get("expected_ref"),
                "base_ref": git.get("base_ref"),
            }
        )
    return result


async def _plan_with_llm(
    issue: Issue,
    config: RepositoryPlanningConfig,
    coding_config: CodingContextConfig,
    codex_config: CodexConfig,
    workspace_path: Path,
) -> RepoPlan:
    planner_codex_config = replace(
        codex_config,
        model=config.plan_model or codex_config.model,
        effort=config.plan_effort,
        turn_timeout_ms=config.plan_timeout_ms,
        summary=None,
        personality=None,
    )

    async def ignore_event(_: dict[str, Any]) -> None:
        return None

    async with CodexAppServerSession(
        planner_codex_config,
        workspace_path,
        tracker_config=None,
        on_event=ignore_event,
    ) as session:
        result = await session.run_turn(_planner_prompt(issue, config, coding_config), capture_agent_text=True)
    data = _parse_json_object(result.agent_message_text)
    return _normalize_plan(issue, data, config, planner="llm", source="llm")


def _planner_prompt(issue: Issue, config: RepositoryPlanningConfig, coding_config: CodingContextConfig) -> str:
    payload = {
        "issue": issue.to_template_data(),
        "repositories": [repo.to_prompt_data() for repo in config.repositories],
        "coding_context": truncate(load_coding_context(coding_config), 30000) if coding_config.enabled else "",
    }
    return (
        "You are Symphony's repository planner for a background coding agent.\n"
        "Pick the repository set the agent is allowed to use for this Linear issue. Many valid tasks span multiple "
        "repos, so return a primary repo plus optional secondary/edit repos and read-only context repos. Prefer the "
        "runtime/product repo that owns the behavior as primary. Do not choose a gateway/config repo merely because "
        "the issue mentions AI, search, prompts, or suggestions if the runtime/UI/provider adapter lives elsewhere. "
        "If the issue requests a specific provider/integration, keep that provider direction in the reason.\n\n"
        "Return only one JSON object, no markdown, no prose, and no tool calls.\n"
        "Schema:\n"
        "{\n"
        '  "coding_task": true,\n'
        '  "primary_repo": {"slug": "owner/name", "reason": "why this repo is the start"},\n'
        '  "secondary_repos": [{"slug": "owner/name", "reason": "why it may need edits", "edit_allowed": true}],\n'
        '  "read_only_context_repos": [{"slug": "owner/name", "reason": "why it is useful context"}],\n'
        '  "confidence": 0.0,\n'
        '  "needs_human": false,\n'
        '  "human_reason": null,\n'
        '  "notes": "short operational note"\n'
        "}\n\n"
        "Rules:\n"
        "- Use only repository slugs listed in the input catalog.\n"
        "- If there is no clear primary repo, set needs_human=true and explain.\n"
        "- If a secondary repo might need edits, include it as secondary_repos with edit_allowed=true.\n"
        "- If a repo is only background material, include it as read_only_context_repos.\n"
        "- If the issue is not a coding/repository task, set coding_task=false and leave repo lists empty.\n\n"
        f"Planner input JSON:\n{json.dumps(payload, sort_keys=True, default=str)}"
    )


def _plan_with_rules(issue: Issue, config: RepositoryPlanningConfig, *, source: str) -> RepoPlan:
    text = f"{issue.identifier}\n{issue.title}\n{issue.description or ''}\n{' '.join(issue.labels)}".lower()
    scored: list[tuple[int, RepositoryConfig, list[str]]] = []
    for repo in config.repositories:
        reasons: list[str] = []
        score = 0
        candidates = {repo.slug.lower(), repo.path_name.lower(), *(alias.lower() for alias in repo.aliases)}
        for candidate in sorted(candidate for candidate in candidates if candidate):
            if candidate in text:
                score += 4 if candidate in {repo.slug.lower(), repo.path_name.lower()} else 2
                reasons.append(candidate)
        if repo.description:
            for word in _keyword_terms(repo.description):
                if word in text:
                    score += 1
                    reasons.append(word)
        if score > 0:
            scored.append((score, repo, reasons[:6]))
    scored.sort(key=lambda row: (-row[0], row[1].slug))
    if not scored:
        return RepoPlan(
            issue_identifier=issue.identifier,
            coding_task=True,
            planner=config.planner,
            source=source,
            needs_human=True,
            human_reason="No configured repository matched the issue text.",
            confidence=0.0,
            created_at=now_utc(),
        )
    top_score, top_repo, top_reasons = scored[0]
    tied = [repo.slug for score, repo, _ in scored if score == top_score]
    needs_human = len(tied) > 1
    primary = _item(top_repo, "primary", f"Rules matched: {', '.join(top_reasons)}")
    secondary = [
        _item(repo, "secondary", f"Rules also matched: {', '.join(reasons)}")
        for score, repo, reasons in scored[1:4]
        if score > 0
    ]
    return RepoPlan(
        issue_identifier=issue.identifier,
        coding_task=True,
        planner=config.planner,
        source=source,
        primary_repo=primary,
        secondary_repos=secondary,
        confidence=min(0.85, max(0.2, top_score / 12)),
        needs_human=needs_human,
        human_reason=f"Rules planner found tied primary repositories: {', '.join(tied)}" if needs_human else None,
        created_at=now_utc(),
    )


def _normalize_plan(
    issue: Issue,
    data: dict[str, Any],
    config: RepositoryPlanningConfig,
    *,
    planner: str,
    source: str,
) -> RepoPlan:
    known = config.repository_by_slug
    unknown: list[str] = []

    def parse_item(raw: Any, role: str) -> RepoPlanItem | None:
        if not isinstance(raw, dict):
            return None
        slug = str(raw.get("slug") or "").strip()
        if not slug:
            return None
        repo = known.get(slug)
        if repo is None:
            unknown.append(slug)
            return None
        edit_allowed = bool(raw.get("edit_allowed", role != "read_only_context")) and role != "read_only_context"
        return _item(repo, role, truncate(str(raw.get("reason") or ""), 500) or None, edit_allowed=edit_allowed)

    coding_task = bool(data.get("coding_task", data.get("is_coding_task", True)))
    primary = parse_item(data.get("primary_repo"), "primary")
    secondary = _dedupe_items(
        [
            item
            for raw in _list_value(data.get("secondary_repos"))
            if (item := parse_item(raw, "secondary")) is not None
        ]
    )
    read_only = _dedupe_items(
        [
            item
            for raw in _list_value(data.get("read_only_context_repos"))
            if (item := parse_item(raw, "read_only_context")) is not None
        ]
    )
    if primary is not None:
        secondary = [item for item in secondary if item.slug != primary.slug]
        read_only = [item for item in read_only if item.slug != primary.slug]
    confidence = _confidence(data.get("confidence"))
    needs_human = bool(data.get("needs_human"))
    human_reason = truncate(str(data.get("human_reason") or ""), 500) or None
    if coding_task and primary is None:
        needs_human = True
        human_reason = human_reason or "Repository planner did not return a primary repo."
    if unknown:
        needs_human = True
        suffix = f"Planner returned unknown repositories: {', '.join(sorted(set(unknown)))}."
        human_reason = f"{human_reason} {suffix}".strip() if human_reason else suffix
    return RepoPlan(
        issue_identifier=issue.identifier,
        coding_task=coding_task,
        planner=planner,
        source=source,
        primary_repo=primary,
        secondary_repos=secondary,
        read_only_context_repos=read_only,
        confidence=confidence,
        needs_human=needs_human,
        human_reason=human_reason,
        notes=truncate(str(data.get("notes") or ""), 500) or None,
        created_at=now_utc(),
    )


def _item(repo: RepositoryConfig, role: str, reason: str | None, *, edit_allowed: bool = True) -> RepoPlanItem:
    if role == "read_only_context":
        edit_allowed = False
    return RepoPlanItem(
        slug=repo.slug,
        role=role,
        reason=reason,
        path_name=sanitize_workspace_key(repo.path_name),
        edit_allowed=edit_allowed,
    )


def _dedupe_items(items: list[RepoPlanItem]) -> list[RepoPlanItem]:
    seen: set[str] = set()
    deduped: list[RepoPlanItem] = []
    for item in items:
        if item.slug in seen:
            continue
        seen.add(item.slug)
        deduped.append(item)
    return deduped


def _list_value(value: Any) -> list[Any]:
    return value if isinstance(value, list) else []


def _confidence(value: Any) -> float | None:
    if isinstance(value, bool) or value is None:
        return None
    try:
        return max(0.0, min(1.0, float(value)))
    except (TypeError, ValueError):
        return None


def _parse_json_object(text: str) -> dict[str, Any]:
    stripped = text.strip()
    if not stripped:
        raise ValueError("repo planner returned empty text")
    try:
        value = json.loads(stripped)
    except json.JSONDecodeError:
        value = json.loads(_extract_json_object(stripped))
    if not isinstance(value, dict):
        raise ValueError("repo planner JSON is not an object")
    return value


def _extract_json_object(text: str) -> str:
    start = text.find("{")
    end = text.rfind("}")
    if start == -1 or end == -1 or end <= start:
        raise ValueError("repo planner output did not contain a JSON object")
    return text[start : end + 1]


def _keyword_terms(text: str) -> list[str]:
    stop = {"the", "and", "for", "with", "from", "that", "this", "repo", "choose", "validation", "local", "path"}
    terms: list[str] = []
    for raw in text.lower().replace("/", " ").replace("-", " ").split():
        word = "".join(ch for ch in raw if ch.isalnum())
        if len(word) >= 5 and word not in stop:
            terms.append(word)
    return terms[:40]
